okhttp/src/main/kotlin/okhttp3/internal/http1/HeadersReader.kt

```diff
@@ -19,15 +19,26 @@ import okhttp3.Headers
 import okio.BufferedSource
 
 /**
- * Parse all headers delimited by "\r\n" until an empty line. This throws if headers exceed 256 KiB.
+ * Parse all headers delimited by "\r\n" until an empty line. This throws if headers exceed `headerLimit`.
  */
-class HeadersReader(val source: BufferedSource) {
-  private var headerLimit = HEADER_LIMIT.toLong()
+class HeadersReader(
+  val source: BufferedSource,
+  private var headerLimit: Long = HEADER_LIMIT
+) {
 
   /** Read a single line counted against the header size limit. */
   fun readLine(): String {
-    val line = source.readUtf8LineStrict(headerLimit)
-    headerLimit -= line.length.toLong()
+    val line = if(headerLimit == HEADER_NO_LIMIT)
+        source.readUtf8LineStrict()
```

Comment:
Given we now have two codepaths here, and it was already confusing. Should we catch and re throw in all cases with context such as remaining header length/"NO_LIMIT"?

Reply:
I'm not sure to understant what you mean here ? If there is no limit and we do `source.readUtf8LineStrict(headerLimit)`, it will throw an exception, and the next `source.readUtf8LineStrict()` will be run on the next line if I understant it well.
So how would you perform the completion of the line in the `catch` clause ?

Reply:
It's probably worth catching and rethrowing and explaining issue so it's clear why

- newline not found within limit 
- newline now found by end of stream

Reply:
Not sure to really understand what you mean. Can you write 1 or 2 line of pseudo code, to make sure I understand it well (no need to make something correct, it does not need to compile either, it's just in order to get an idea of what you mean)

Reply:
I'm mainly concerned about the confusion with this limit and the limit used by Okio.

https://github.com/square/okio/blob/597a65527c987d9f221dd5607c49e49f934df2de/okio/src/jvmTest/kotlin/okio/ReadUtf8LineTest.kt#L66

```
  throw EOFException(
    "\\n not found: limit=" + minOf(buffer.size, limit) +
      " content=" + data.readByteString().hex() + '…'.toString(),
  )
```

So if I've set the header length limit to 1024, but I might get 0, 10, or 1024 as the limit in the exception message. I'd be confused if I saw a different limit here.

Maybe something like

```
    val line = try {
      if (headerLimit == HEADER_NO_LIMIT) source.readUtf8LineStrict()
      else source.readUtf8LineStrict(headerLimit)
    } catch (eofe: EOFException) {
      throw EOFException(buildString {
        append("Could not read a complete header")
        if (headerLimit != HEADER_NO_LIMIT) {
          append(" within $headerLimit bytes")
        }
      }).apply {
        initCause(eofe)
      }
    }
``

But I'm open to suggestions.